transport/grpc/integration_test.go

```diff
@@ -369,6 +371,192 @@ func TestCustomContextDial(t *testing.T) {
+			want: 0,
+		},
+		{
+			msg: "fail compression",
+			testEnvOptions: testEnvOptions{
+				DialOptions: []DialOption{Compressor(compressors[1].Name())},
+			},
+			// We should have the following chain of events:
+			// - compress request
+			want:    1,
+			wantErr: "code:internal message:grpc: error while compressing: assert.AnError general error for testing",
+		},
+		{
+			msg: "fail decompression",
+			testEnvOptions: testEnvOptions{
+				DialOptions: []DialOption{Compressor(compressors[2].Name())},
+			},
+			// We should have the following chain of events:
+			// - compress request
+			// - decompress request
+			want:    2,
+			wantErr: "code:internal message:grpc: failed to decompress the received message assert.AnError general error for testing",
+		},
+		{
+			msg: "gzip",
+			testEnvOptions: testEnvOptions{
+				DialOptions: []DialOption{Compressor(compressors[3].Name())},
+			},
+			// We should have the following chain of events:
+			// - compress request
+			// - decompress request
+			// - compress response
+			// - decompress response
+			// - TODO: what else? there seem to be more
+			want: 3,
+			// TODO: something is wrong here
+			wantErr: "code:internal message:grpc: failed to decompress the received message unexpected EOF",
+		},
+	}
+
+	for i := range tests {
+		tt := tests[i]
+		value := strings.Repeat("a", 32768)
+		tt.do(t, func(t *testing.T, e *testEnv) {
+			err := e.SetValueYARPC(context.Background(), "foo", value)
+			if tt.wantErr != "" {
+				assert.Error(t, err)
+				assert.EqualError(t, err, tt.wantErr)
+			} else if assert.NoError(t, err) {
+				getValue, err := e.GetValueYARPC(context.Background(), "foo")
+				assert.NoError(t, err)
+				assert.Equal(t, value, getValue)
+			}
+
+			if tt.want < 0 {
+				return
+			}
+			comp, ok := compressors[tt.want].(*testCompressor)
+			assert.True(t, ok)
+			assert.Equal(t, metrics[tt.want], comp.metrics)
+		})
+	}
+}
+
+type testCompressor struct {
+	name       string
+	metrics    compressionMetrics
+	comperr    error
+	decomperr  error
+	enableGZip bool
+}
+
+func newCompressor(name string) transport.Compressor {
+	comp := testCompressor{
+		name:    name,
+		metrics: compressionMetrics{},
+	}
+
+	switch name {
+	case "fail-comp":
+		comp.comperr = assert.AnError
+	case "fail-decomp":
+		comp.decomperr = assert.AnError
+	case "gzip":
+		comp.enableGZip = true
+	}
+	return &comp
+}
+
+type compressionMetrics struct {
+	Compress     int
+	Decompress   int
+	BytesWritten int
+}
+
+func (c *testCompressor) Name() string { return c.name }
+
+func (c *testCompressor) Compress(w io.Writer) (io.WriteCloser, error) {
+	c.metrics.Compress++
+
+	if c.enableGZip {
+		w = gzip.NewWriter(w)
+	}
+
+	return &meteredWriter{
+		WriteCloser: nopCloser(w),
+		metrics:     &c.metrics,
+	}, c.comperr
+}
+
+// Decompress maybe should return io.ReadCloser? because it is rather weird why you have this.
+func (c *testCompressor) Decompress(r io.Reader) (io.Reader, error) {
```

Comment:
The signature of this is strange. I would expect something else.

Reply:
If you see the comment above (in the code, next to the function), I am wondering if this Decompress signature should not return `io.ReadCloser`, which would allow the `Close` to be called just like in the [gzip#Reader.Close](https://godoc.org/compress/gzip#Reader.Close). However, this may be irrelevant because not all readers may be `ReadClosers`, like in [snappy's case](https://godoc.org/github.com/golang/snappy#Reader).